WordPress/wordpress-develop
# Block Hooks API: Update block-template-utils for first-last child template parts

Currently you are unable to insert hooked blocks in `first_child` and `last_child` positions of template parts. This is because, similarly to the Navigation block the anchor block (`core/navigation` / `core/template-part`) and its children are detached in the markup from one and other.

To solve this issue we have approached it in a similar way by mocking the anchor block in such cases for the purpose of running the block hooks algorithm against it and extracting any related metadata (e.g. `ignoredHookedBlocks`) which are then stored in postmeta.

Trac ticket: https://core.trac.wordpress.org/ticket/60854

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### Testing instructions

1. Use both methods below to hook the `core/loginout` block as a `first_child` into template parts.
2. Ensure it renders as expected for unmodified templates.
3. Move/remove the block from the template part and ensure its not readded.
4. Remove the `_wp_ignored_hooked_blocks` data from `wp_postmeta` from your site and perform steps 1-3 again for modified templates.

`functions.php`
```php
function register_logout_block_as_hooked_block( $hooked_blocks, $position, $anchor_block, $context ) {
	if ( 'core/template-part' === $anchor_block && 'first_child' === $position ) {
		$hooked_blocks[] = 'core/loginout';
	}

	return $hooked_blocks;
}

add_filter( 'hooked_block_types', 'register_logout_block_as_hooked_block', 10, 4 );
```

`block.json`
```json
"blockHooks": {
    "core/template-part": "lastChild"
}
```
